lib/composite.ex

```diff
@@ -387,6 +406,21 @@ defmodule Composite do
     {query, MapSet.union(loaded_deps, deps_to_load)}
   end
 
+  defp maybe_raise_on_unknown_params(%__MODULE__{strict: true} = composite) do
+    diff =
+      MapSet.difference(
+        MapSet.new(Map.keys(composite.params)),
+        MapSet.new(composite.param_definitions |> Enum.map(&elem(&1, 0)) |> List.flatten())
```

Comment:
It shouldn't be flattened at the end. This list means `path` to the value.
Example from docs:
```elixir
params = %{filter: %{name: "John"}}

User
|> Composite.new(params)
|> Composite.param([:filter, :name], &where(&1, name: ^&2))
```
This code will return `MapSet.new([:name, :filter])` for the struct defined above.

Reply:
It would be great to handle nested containers as well. I suggest updating your function to this
```elixir
defp maybe_raise_on_unknown_params(composite) do
  paths = Enum.map(composite.param_definitions, &elem(&1, 0))
  maybe_raise_on_unknown_params(composite.params, paths, [])
end

defp maybe_raise_on_unknown_params(params, paths, current_path) do
  if (is_map(params) and not is_struct(params)) or Keyword.keyword?(params) do
    paths = Enum.group_by(paths, &hd/1, &tl/1)

    absent_paths =
      Enum.flat_map(params, fn {key, value} ->
        case Enum.find(paths, fn {path_key, _subpaths} ->
               path_key == key
             end) do
          nil ->
            [current_path ++ [key]]

          {^key, subpaths} ->
            maybe_raise_on_unknown_params(value, subpaths, current_path ++ [key])
            []
        end
      end)

    if absent_paths != [] do
      raise ArgumentError,
            "Unknown parameters found under the following paths: #{Enum.map_join(absent_paths, ", ", &inspect/1)}"
    end
  end

  :noop
end
```


As a final step, could you also add a test? :slightly_smiling_face: